### DISCUSSION TITLE
How to extrude 2D grains to 3D

### DISCUSSION THREAD

**Discussion by Yuanguo1994** [CsatiZoltan/CristalX#68] [Q&A]
Hi, CsatiZoltan

Thanks for you providing such helpful software. I have a problem in extruding a 2D grain microstructure to 3D. Could you please tell me how could you do that? or what kind of software you use to do that. Could it be achieved by CristalX?

**Comment by CsatiZoltan** [comment]
Hi @Yuanguo1994 !

I was working on it a few months ago, so I can answer to this question in great details. I plan to write a tutorial about it within a few weeks.
Basically, there are two options (what I know about):
1. Extrude the **geometry** of the grains
   Once you have an [explicit representation](https://cristalx.readthedocs.io/en/latest/example_application.html#From-image-to-geometry) of a grain as a B-spline surface, it can be extruded by various CAD software. I successfully used [Salome](https://docs.salome-platform.org/latest/gui/GEOM/create_extrusion_page.html) for this purpose. If your goal is to obtain a 3D geometry, this is the way to go.
2. Extrude the **mesh** representation of the grains
   If you want to mesh the extruded grains, you can simply mesh the 2D microstructure first, and then extrude the mesh. This is also possible in [Salome](https://docs.salome-platform.org/latest/gui/SMESH/extrusion.html). Note, however, that this will most probably yield a mesh with bad quality elements. If your final goal is to have a good quality mesh, it is recommended to first extrude the 2D geometry to 3D (see the previous point) and then mesh the 3D domain.

Extruding the 2D microstructure or its mesh is not yet supported in *CristalX*. The reason is that this seemingly simple task more complicated than it seems. The extrusion itself is easy, but one has to play with the tolerances in Salome to obtain a topologically valid 3D geometry. I am not sure that this is possible to fully automate, therefore it is best done in Salome (or other CAD software of your preference).

In the upcoming weeks, I will greatly expand the documentation, and push new functionalities to the GitHub repository.

**Reply by Yuanguo1994** [reply]
Thank you very much! I will try to do that. Also, I am looking forward to your tutorial. Could you please remind me or just leave a comment on the discussion, once you release the tutorial?

**Reply by CsatiZoltan** [reply]
Yes, I will let you know in this discussion.
